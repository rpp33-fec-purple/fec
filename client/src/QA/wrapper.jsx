import React from 'react';

class QAndA extends React.Component {
  constructor(props) {
    super(props);
  }

  render() {
    return (
      <h2>Questions and Answers Component</h2>
    )
  }
}

export default QAndA;