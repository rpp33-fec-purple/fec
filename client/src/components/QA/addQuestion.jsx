import React from 'react';
import QuestionModal from './Modals/addQuestionModal.jsx';
import styled from 'styled-components';
import GlobalStyle from './globalStyles/globalStyles.jsx'

const Container = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  // height: 10vh;
`

const Button = styled.button`
  min-width: 80px;
  padding: 16px 32px;
  border-radius: 4px;
  border: none;
  background: #141414;
  color: #fff;
  font-size: 18px;
  cursor: pointer;
`

class AddQuestion extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isModalShowing: false
    }
    this.addAQ = this.addAQ.bind(this);
    this.changeModalVisibilityState = this.changeModalVisibilityState.bind(this);
  }

  changeModalVisibilityState() {
    this.setState({
      isModalShowing: !this.state.isModalShowing
    }, ()=> {
      console.log('is modal showing in changestate func', this.state.isModalShowing)
    })
  }

  addAQ(){
    console.log('Add a question clicked!')
  }


  render() {
    return (
      <Container>
        <Button onClick={this.changeModalVisibilityState}>ADD A QUESTION +</Button>
        <QuestionModal productName={this.props.productName} isModalShowing={this.state.isModalShowing} changeModalState={this.changeModalVisibilityState}/>
        <GlobalStyle/>
      </Container>
    )
  }
}

export default AddQuestion;




// const Button = styled.button`
//   min-width: 100px;
//   padding: 16px 32px;
//   border-radius: 4px;
//   border: none;
//   background: #141414;
//   color: #fff;
//   font-size: 24px;
//   cursor: pointer;
// `