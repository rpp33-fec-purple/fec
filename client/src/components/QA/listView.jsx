import React from 'react';
import $ from 'jquery';
import QuestionListEntry from './questionListEntry.jsx';
import styled from 'styled-components';


const Container = styled.div`
  display: flex;
  // justify-content: center;
  // align-items: center;
  // height: 70%;
  // box-sizing: border-box;
  // margin: 0;
  // padding: 0;
  // font-family: 'Arial', sans-serif;
`

const ScrollableList = styled.div`
      margin: 0 auto;
      // height: 250px;
      min-height: 100px;
      max-height: 550px;
      // max-height:75%;
      width: 800px;
      overflow: auto;
      border: 1px solid black;
      // display: flex;
      // justify-content: center;
      align-items: center;
  `;

class ListView extends React.Component {
  constructor(props) {
    super(props);
    this.state= {
      inDisplay: 2,
      noMoreToDisplay: false,
      updatedQandAList: []
    }
    this.seeMoreQuestions = this.seeMoreQuestions.bind(this);
  }

  seeMoreQuestions() {
    $.ajax({
      url: `http://localhost:3000/qa/questions`,
      data: {
        product_id: this.props.productID,
        page: 1,
        count: 100
      },
      method: 'GET',
      success: (data) => {
        console.log('data in client', data);

        var totalNumOfQs = data.results.length;

        if (totalNumOfQs > this.state.inDisplay){
          if (this.state.inDisplay + 2 >= totalNumOfQs) {
            this.setState({
              inDisplay: this.state.inDisplay + 2,
              noMoreToDisplay: true,
              updatedQandAList: data.results
            })
          } else {
            this.setState({
              inDisplay: this.state.inDisplay + 2,
              updatedQandAList: data.results
            })
          }
        }

      },
      error: (err) => {
        console.log('Error with GET request:', err);
      }
    });


    // var totalNumOfQs = this.props.qAndAList.length;

    // if (totalNumOfQs > this.state.inDisplay){
    //   if (this.state.inDisplay + 2 >= totalNumOfQs) {
    //     this.setState({
    //       inDisplay: this.state.inDisplay + 2,
    //       noMoreToDisplay: true
    //     })
    //   } else {
    //     this.setState({
    //       inDisplay: this.state.inDisplay + 2
    //     })
    //   }
    // }

  }

  render () {

    if (this.state.updatedQandAList.length > 0) {
      var questionsList = this.state.updatedQandAList.map(qa => {

        return (
          <div key={qa.question_id}>
            <QuestionListEntry rerenderQandAs={this.props.rerenderQandAs} qACombo={qa} productName={this.props.productName}/>
          </div>
        )
      })

    } else {
      var questionsList = this.props.qAndAList.map(qa => {

        return (
          <div key={qa.question_id}>
            <QuestionListEntry rerenderQandAs={this.props.rerenderQandAs} qACombo={qa} productName={this.props.productName}/>
          </div>
        )
      })
    }


    const questionsInView = questionsList.slice(0, this.state.inDisplay);
    let moreQuestionsButton = this.state.noMoreToDisplay ? <div/> : <button onClick={this.seeMoreQuestions.bind(this)}>MORE ANSWERED QUESTIONS</button>

    return(
      <div id='qaView'>
        <Container>
          <ScrollableList>{questionsInView}</ScrollableList>
        </Container>
        <div>{moreQuestionsButton}</div>
      </div>
    )
  }
}

export default ListView;

////////////////////////////////////////////////////////

////////////////////////////////////////////////////////

////////////////////////////////////////////////////////

////////////////////////////////////////////////////////


// import React from 'react';
// import $ from 'jquery';


// class ListView extends React.Component {
//   constructor(props) {
//     super(props);
//     this.state= {
//       inDisplay: 2,
//       noMoreToDisplay: false
//     }
//     this.seeMoreQuestions = this.seeMoreQuestions.bind(this);
//     this.seeMoreAnswers = this.seeMoreAnswers.bind(this);
//   }

//   seeMoreQuestions() {
//     var totalNumOfQs = this.props.qAndAList.length;
//     console.log('list', this.props.qAndAList)



//     if (totalNumOfQs > this.state.inDisplay){
//       if (this.state.inDisplay + 2 >= totalNumOfQs) {
//         this.setState({
//           inDisplay: this.state.inDisplay + 2,
//           noMoreToDisplay: true
//         })
//       } else {
//         this.setState({
//           inDisplay: this.state.inDisplay + 2
//         })
//       }
//     }
//   }

//   seeMoreAnswers() {
//     // make api call here to get all the answers?
//   }

//   render (){
//     const questionsList = this.props.qAndAList.map(qa => {
//       var answers = [];
//       for (var key in qa.answers) {
//         answers.push(qa.answers[key])
//       }
//       var answersDiv = answers.map(a =>
//         <div key={a.id}>
//           A: {a.body}
//           <br></br> by {a.answerer_name}, {a.date} | Helpful? Yes ({a.helpfulness}) | Report
//         </div>
//       )
//       return (
//         <li key={qa.question_id}>
//         Q: {qa.question_body}
//         <br></br> Helpful? Yes ({qa.question_helpfulness}) | Add Answer
//         <div>
//           {answersDiv}
//           <button onClick={this.seeMoreAnswers.bind(this)}>SEE MORE ANSWERS</button>
//         </div>
//       </li>
//       )
//     })

//     const questionsInView = questionsList.slice(0, this.state.inDisplay);
//     let moreQuestionsButton = this.state.noMoreToDisplay ? <div/> : <button onClick={this.seeMoreQuestions.bind(this)}>MORE ANSWERED QUESTIONS</button>

//     return (
//       <div id='qaView'>
//         <ul>{questionsInView}</ul>
//         <div>{moreQuestionsButton}</div>
//       </div>
//     )
//   }
// }

// export default ListView;