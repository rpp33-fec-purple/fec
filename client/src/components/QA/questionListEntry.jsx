import React from 'react';
import $ from 'jquery';
import styled from 'styled-components';
import AnswerModal from './Modals/addAnswerModal.jsx';
import ImageModal from './Modals/imageModal.jsx';
import baseUrl from './../../../../config.js';

const Container = styled.div`
  display: flex;
  justify-content: left;
  padding: 4px;
  width: 780px;
  position: inherit;
  .moreAnswersButton {
  padding: .2rem 2rem;
  }
`;


  const AnswersScrollableList = styled.div`
  margin: 0 auto;
  max-height: 400px;
  width: 100%;
  overflow: auto;
  align-items: center;
`;


  const ImageThumbnail = styled.img`
    border: 2px solid #C0C0C0;
    padding: 5px;
    margin-right: 5px;
    margin-left: 5px;
    margin-top: 2.5px;
    margin-bottom: 2.5px;
    height: 60px;
    cursor: pointer;
  `;

  const QuestionBox = styled.div`
    .questionBox {
      display: flex;
      flex-wrap: nowrap;
      justify-content: space-between;
      width: 780px;
      padding-bottom: .2rem;
    }

    .questionTitle {
      width: 2rem;
    }

    .push {
        margin-left: auto;
        justify-content: right;
        font-size: 11.5px;
        color: #808080;
    }
    .question {
      width: 28.12em;
      flex-wrap: wrap;
      overflow-wrap: anywhere;
    }

    .report {
      padding-left: .5rem;
      padding-right: .5rem;
      cursor: pointer;

    }
    .answer {
      padding-left: .5rem;
      cursor: pointer;
    }

    .questionHelpfulSpan {
      padding-right: .5rem;
    }

    .helpfulLink {
      cursor: pointer;
    }
  `;

    const AnswersContainer = styled.div`
      .answersContainer {
        display: flex;
        flex-wrap: nowrap;
        justify-content: flex-start;
        width: 700px;
      }

      .answersTitle {
        width: 2rem;
        padding-top: .5rem;
        padding-bottom: .2rem;
      }

      .answersInView {
        padding-left: 1px;
        width: 400px;
      }

      .answerInfo {
        font-size: 13.5px;
        padding-top: .5rem;
        padding-bottom: .5rem;
      }
      .markAnswerHelpful {
        cursor: pointer;

      }

      .reportAnswer {
        cursor: pointer;
        padding-left: .5rem;
      }

      .answererInfoAndLinks {
        font-size: 11.5px;
        color: #808080;
        padding-top: .5rem;
      }

      .helpfulSpan {
        padding-left: .5rem;
        padding-right: .5rem;
      }

      .answererInfoSpan {
        padding-right: .5rem;
      }
    `;


  const AnswersButton = styled.div`

  .button-forAnswers {
    appearance: none;
    background-color: #FFFFFF;
    border-width: 0;
    box-sizing: border-box;
    color: #000000;
    cursor: pointer;
    display: inline-block;
    font-family: Clarkson,Helvetica,sans-serif;
    font-size: 12px;
    font-weight: 500;
    line-height: .7em;
    opacity: 1;
    outline: 0;
    padding: .8em 1.6em;
    position: relative;
    text-align: center;
    text-rendering: geometricprecision;
    transition: opacity 300ms cubic-bezier(.694, 0, 0.335, 1),background-color 100ms cubic-bezier(.694, 0, 0.335, 1),color 100ms cubic-bezier(.694, 0, 0.335, 1);
    touch-action: manipulation;
    vertical-align: baseline;
    white-space: nowrap;
  }

  .button-forAnswers:before {
    animation: opacityFallbackOut .5s step-end forwards;
    backface-visibility: hidden;
    background-color: #EBEBEB;
    clip-path: polygon(-1% 0, 0 0, -25% 100%, -1% 100%);
    content: "";
    height: 100%;
    left: 0;
    position: absolute;
    top: 0;
    transform: translateZ(0);
    transition: clip-path .5s cubic-bezier(.165, 0.84, 0.44, 1), -webkit-clip-path .5s cubic-bezier(.165, 0.84, 0.44, 1);
    width: 100%;
  }

  .button-forAnswers:hover:before {
    animation: opacityFallbackIn 0s step-start forwards;
    clip-path: polygon(0 0, 101% 0, 101% 101%, 0 101%);
  }

  .button-forAnswers:after {
    background-color: #FFFFFF;
  }

  .button-forAnswers span {
    z-index: 1;
    position: relative;
  }
  `;

class QuestionListEntry extends React.Component {
  constructor(props) {
    super(props);
    this.state= {
      answersInDisplay: 2,
      displayingAll: false,
      isModalShowing: false,
      updatedAnswersList: [],
      answersMarkedHelpful: [],
      isImageModalShowing: false
    }
    this.seeMoreAnswers = this.seeMoreAnswers.bind(this);
    this.collapseAnswers = this.collapseAnswers.bind(this);
    this.sortAnswers = this.sortAnswers.bind(this);
    this.formatDate = this.formatDate.bind(this);
    this.changeModalVisibilityState = this.changeModalVisibilityState.bind(this);
    this.markQuestionHelpful = this.markQuestionHelpful.bind(this);
    this.reportQuestion = this.reportQuestion.bind(this);
    this.markAnswerHelpful = this.markAnswerHelpful.bind(this);
    this.reportAnswer = this.reportAnswer.bind(this);
    this.disableButton = this.disableButton.bind(this);
    this.changeImageModalVisibilityState = this.changeImageModalVisibilityState.bind(this);

    var newAnswerAdded = false;
    var answerMarkedHelpful = false;
    var seeMoreAnswersClicked = false;
    var collapseAnswersClicked = false;
    var questionMarkedHelpful = false;
    var imageThumbnailClicked = '';
  }


  changeImageModalVisibilityState() {
    console.log('image clicked', event.target.src)
    this.imageThumbnailClicked = event.target.src;
    this.setState({
      isImageModalShowing: !this.state.isImageModalShowing
    })
  }

  changeModalVisibilityState(answerAdded) {

    if (answerAdded) {
      $.ajax({
        url: `${baseUrl}/qa/questions/${this.props.qACombo.question_id}/answers`,
        data: {
          question_id: this.props.qACombo.question_id,
          page: 1,
          count: 100
        },
        method: 'GET',
        success: (data) => {
          this.newAnswerAdded = true;

          this.setState({
            isModalShowing: !this.state.isModalShowing,
            updatedAnswersList: data.results
          })
        },
        error: (err) => {
          console.log('Error with POST request:', err);
          this.setState({
            isModalShowing: !this.state.isModalShowing
          })
        }
      })
    } else {
    this.setState({
      isModalShowing: !this.state.isModalShowing
    }, ()=> {
      console.log('is modal showing in changestate func', this.state.isModalShowing)
    })

    }
  }

  collapseAnswers() {
    $.ajax({
      url: `${baseUrl}/qa/questions/${this.props.qACombo.question_id}/answers`,
      data: {
        question_id: this.props.qACombo.question_id,
        page: 1,
        count: 100
      },
      method: 'GET',
      success: (data) => {
        var totalNumOfAs = data.results.length;
        this.collapseAnswersClicked = true;
        this.setState({
          answersInDisplay: 2,
          displayingAll: false,
          updatedAnswersList: data.results
        })
      },
      error: (err) => {
        console.log('Error with GET request:', err);
      }
    })
  }

  seeMoreAnswers() {
    $.ajax({
      url: `${baseUrl}/qa/questions/${this.props.qACombo.question_id}/answers`,
      data: {
        question_id: this.props.qACombo.question_id,
        page: 1,
        count: 100
      },
      method: 'GET',
      success: (data) => {
        // console.log('these are the answers', data.results)
        var totalNumOfAs = data.results.length;
        this.seeMoreAnswersClicked = true;
        if (totalNumOfAs > this.state.answersInDisplay) {
          this.setState({
            answersInDisplay: totalNumOfAs,
            displayingAll: true,
            updatedAnswersList: data.results
          })
        }
      },
      error: (err) => {
        console.log('Error with GET request:', err);
      }
    })
  }


  sortAnswers(answers) {
    var sorted = answers.sort((a, b) => b.helpfulness - a.helpfulness);
    var sellerFirst = []
    for (var i = 0; i < sorted.length; i++) {
      if (sorted[i].answerer_name === 'Seller') {
        sellerFirst.push(sorted[i]);
        sorted.splice(i, 1);
      }
    }
    sellerFirst = sellerFirst.concat(sorted);
    return sellerFirst;
  }

  formatDate(isoDate) {
    const months = [
      'January',
      'February',
      'March',
      'April',
      'May',
      'June',
      'July',
      'August',
      'September',
      'October',
      'November',
      'December'
    ]
    const date = new Date(isoDate);
    const year = date.getFullYear();
    const day = date.getDate();
    const monthIndex = date.getMonth();
    const monthName = months[monthIndex];

    return `${monthName} ${day}, ${year}`
  }

  markQuestionHelpful(questionID) {
    console.log('event target id', event.target)
    this.disableButton(event.target.id);
    var buttonID = event.target.id;

    $.ajax({
      url: `${baseUrl}/qa/questions/${questionID}/helpful`,
      data: {
        question_id: questionID
      },
      method: 'PUT',
      success: (data) => {
        this.props.rerenderQandAs()

      },
      error: (err) => {
        console.log('Error with POST request:', err);
      }
    })
  }

  reportQuestion(questionID) {
    console.log('event target id', event.target)
    this.disableButton(event.target.id);
    var buttonID = event.target.id;

    $.ajax({
      url: `${baseUrl}/qa/questions/${questionID}/report`,
      data: {
        question_id: questionID
      },
      method: 'PUT',
      success: (data) => {
        $(`#${buttonID}`).html('Reported');
      },
      error: (err) => {
        console.log('Error with POST request:', err);
      }
    })
  }

  markAnswerHelpful(answerID) {
    var buttonID = event.target.id;

    $.ajax({
      url: `${baseUrl}/qa/answers/${answerID}/helpful`,
      data: {
        answer_id: answerID
      },
      method: 'PUT',
      success: (data) => {
        $.ajax({
          url: `${baseUrl}/qa/questions/${this.props.qACombo.question_id}/answers`,
          data: {
            question_id: this.props.qACombo.question_id,
            page: 1,
            count: 100
          },
          method: 'GET',
          success: (data) => {
            this.answerMarkedHelpful = true;
            this.disableButton(buttonID);
            console.log('Successfully marking answer as helpful')
            this.setState({
              updatedAnswersList: data.results
            })
          },
          error: (err) => {
            console.log('Error with GET request:', err);
          }
        })
      },
      error: (err) => {
        console.log('Error with answer PUT request when marking helpful:', err);
      }
    })
  }

  reportAnswer(answerID) {
    console.log('event target id', event.target)

    this.disableButton(event.target.id);
    var buttonID = event.target.id;

    $.ajax({
      url: `${baseUrl}/qa/answers/${answerID}/report`,
      data: {
        answer_id: answerID
      },
      method: 'PUT',
      success: (data) => {
        $(`#${buttonID}`).html('Reported');
      },
      error: (err) => {
        console.log('Error with POST request:', err);
      }
    })
  }

  disableButton(buttonID) {
    document.getElementById(buttonID).style.pointerEvents = 'none';
  }

  render(){
    var answers = [];
    var answersDiv;

    if (this.newAnswerAdded || this.answerMarkedHelpful || this.seeMoreAnswersClicked || this.collapseAnswersClicked) {
      this.newAnswerAdded = false;
      this.answerMarkedHelpful = false;
      this.seeMoreAnswersClicked = false;
      this.collapseAnswersClicked = false;
      var sorted = this.sortAnswers(this.state.updatedAnswersList);

      answersDiv = sorted.map(a => {
        if (a.answerer_name === 'Seller') {
          var seller = <strong key={a.answerer_name}>{a.answerer_name}</strong>
        }

        if (a.photos.length === 0) {
          var photosDiv = <></>;
        } else {
          var photosDiv = a.photos.map(photo => {
            return (
              <ImageThumbnail className='photoThumbnail' key={photo.url} src={photo.url} width='50' onClick={() => { this.changeImageModalVisibilityState() }}></ImageThumbnail>
            )
          })
        }

        return (
          <div key={a.answer_id} className='answerInfo'>
            {a.body}
          <br></br>
          <>{photosDiv}</>

          <div className='answererInfoAndLinks'>
            <p>
              <span className='answererInfoSpan'>by {seller || a.answerer_name}, {this.formatDate(a.date)}</span > |
              <span className='helpfulSpan'> Helpful? <u className='markAnswerHelpful' id={a.answer_id + 'helpful'} onClick={() => { this.markAnswerHelpful(a.answer_id) }}> Yes</u>  ({a.helpfulness})</span> |
              <u className='reportAnswer' id={a.answer_id + 'report'} onClick={ () => { this.reportAnswer(a.answer_id) }}> Report</u>
             </p>
          </div>
        </div>
        )
      }
      )
    } else {
      var QandA = this.props.qACombo;
      for (var key in QandA.answers) {
        answers.push(QandA.answers[key])
      }
      var sorted = this.sortAnswers(answers);

      answersDiv = sorted.map(a => {
        if (a.answerer_name === 'Seller') {
          var seller = <strong key={a.answerer_name}>{a.answerer_name}</strong>
        }


        if (a.photos.length === 0) {
          var photosDiv = <></>;
        } else {
          var photosDiv = a.photos.map((photoURL, i) => {
            return (
              <ImageThumbnail className='photoThumbnail' key={i} src={photoURL} width='50' onClick={() => { this.changeImageModalVisibilityState(photoURL) }}></ImageThumbnail>
            )
          })
        }

        return (
          <div key={a.id} className='answerInfo'>
            {a.body}
            <br></br>
            <>{photosDiv}</>
            <div className='answererInfoAndLinks'>
              <p>
                <span className='answererInfoSpan'>by {seller || a.answerer_name}, {this.formatDate(a.date)}</span> |
                <span className='helpfulSpan'> Helpful? <u className='markAnswerHelpful' id={a.id + 'helpful'} onClick={() => { this.markAnswerHelpful(a.id) }}> Yes</u>  ({a.helpfulness})</span> |
                <u className='reportAnswer' id={a.id + 'report'} onClick={ () => { this.reportAnswer(a.id) }}> Report</u>
              </p>
            </div>
          </div>
        )
      }
      )
    }

    if (this.state.displayingAll) {
      var answersInView = <AnswersScrollableList>{answersDiv.slice(0, this.state.answersInDisplay)}</AnswersScrollableList>
    } else {
      var answersInView = answersDiv.slice(0, this.state.answersInDisplay);
    }

    if (answersDiv.length > 2) {
      var moreAnswersButton = this.state.displayingAll ? <AnswersButton><button onClick={this.collapseAnswers} className="button-forAnswers" role="button"><span className="text">COLLAPSE ANSWERS</span></button></AnswersButton> : <AnswersButton><button onClick={this.seeMoreAnswers.bind(this)} className="button-forAnswers" role="button"><span className="text">SEE MORE ANSWERS</span></button> </AnswersButton>

    } else {
      var moreAnswersButton = <div/>
    }

    var answersTitle = answersDiv.length > 0 ? <strong>A:</strong> : <></>

    return (
      <Container>

        <div key={this.props.qACombo.question_id} >
          <QuestionBox className='questionBox'>
            <div className='questionBox'>
              <div className='questionTitle'>
              <strong>Q: </strong>

              </div>
              <div className='question'>
                <strong>{this.props.qACombo.question_body}</strong>
              </div>
              <div className='push'>
                <span className='questionHelpfulSpan'>Helpful? <u className='helpfulLink' id={this.props.qACombo.question_id + 'helpful'} onClick={ () => { this.markQuestionHelpful(this.props.qACombo.question_id) } }> Yes</u>  ({this.props.qACombo.question_helpfulness}) </span>
                |  <u className='report' id={this.props.qACombo.question_id + 'report'} onClick={() => { this.reportQuestion(this.props.qACombo.question_id) }}> Report</u>  |  <u className='answer' onClick={() => { this.changeModalVisibilityState(false) }}> Add Answer</u>
              </div>
            </div>
          </QuestionBox>
          <div>
           <AnswerModal questionID={this.props.qACombo.question_id} question={this.props.qACombo.question_body} productName={this.props.productName} isModalShowing={this.state.isModalShowing} changeModalState={this.changeModalVisibilityState}></AnswerModal>
           <ImageModal imageSrc={this.imageThumbnailClicked} isImageModalShowing={this.state.isImageModalShowing} changeImageModalState={this.changeImageModalVisibilityState}></ImageModal>
          <div>
            <AnswersContainer>
              <div className='answersContainer'>
                <div className='answersTitle'>{answersTitle}</div>
                <div className='answersInView'>{answersInView}</div>
              </div>
            </AnswersContainer>
            <div className='moreAnswersButton'>
              {moreAnswersButton}
            </div>
          </div>

          </div>
        </div>
      </Container>

    )
  }

}

export default QuestionListEntry;


