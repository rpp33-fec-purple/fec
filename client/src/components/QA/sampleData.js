var sampleProductQuestions = {
  "product_id": "5",
  "results":
  [
    {
      "question_id": 37,
      "question_body": "Why is this product cheaper here than other sites?",
      "question_date": "2018-10-18T00:00:00.000Z",
      "asker_name": "williamsmith",
      "question_helpfulness": 4,
      "reported": false,
      "answers": {
        68: {
          "id": 68,
          "body": "We are selling it here without any markup from the middleman!",
          "date": "2018-08-18T00:00:00.000Z",
          "answerer_name": "Seller",
          "helpfulness": 4,
          "photos": []
        }
      }
    },
    {
      "question_id": 38,
      "question_body": "How long does it last?",
      "question_date": "2019-06-28T00:00:00.000Z",
      "asker_name": "funnygirl",
      "question_helpfulness": 2,
      "reported": false,
      "answers": {
        70: {
          "id": 70,
          "body": "Some of the seams started splitting the first time I wore it!",
          "date": "2019-11-28T00:00:00.000Z",
          "answerer_name": "sillyguy",
          "helpfulness": 6,
          "photos": [],
        },
        78: {
          "id": 78,
          "body": "9 lives",
          "date": "2019-11-12T00:00:00.000Z",
          "answerer_name": "iluvdogz",
          "helpfulness": 31,
          "photos": [],
        }
      }
    },
    {
      "question_id": 88,
      "question_body": "What day is it tomorrow?",
      "question_date": "2078-10-18T00:00:00.000Z",
      "asker_name": "rudy",
      "question_helpfulness": 199,
      "reported": false,
      "answers": {
        89: {
          "id": 89,
          "body": "wuttttt!",
          "date": "2018-08-18T00:00:00.000Z",
          "answerer_name": "Seller",
          "helpfulness": 4,
          "photos": []
        }
      }
    },
    {
      "question_id": 9999,
      "question_body": "What day was it yesterday?",
      "question_date": "2078-10-18T00:00:00.000Z",
      "asker_name": "carlos",
      "question_helpfulness": 7,
      "reported": false,
      "answers": {
        55: {
          "id": 55,
          "body": "stop!",
          "date": "2018-08-18T00:00:00.000Z",
          "answerer_name": "tom",
          "helpfulness": 0,
          "photos": []
        }
      }
    },
    {
      "question_id": 3,
      "question_body": "What year is it?",
      "question_date": "3031-10-18T00:00:00.000Z",
      "asker_name": "jp",
      "question_helpfulness": 50,
      "reported": false,
      "answers": {
        9: {
          "id": 9,
          "body": "give me food",
          "date": "2018-08-18T00:00:00.000Z",
          "answerer_name": "angie",
          "helpfulness": 876,
          "photos": []
        }
      }
    }
  ]
};

var answersList = {
    "question": "1",
    "page": 0,
    "count": 5,
    "results": [
      {
        "answer_id": 8,
        "body": "What a great question!",
        "date": "2018-01-04T00:00:00.000Z",
        "answerer_name": "metslover",
        "helpfulness": 8,
        "photos": [],
      },
      {
        "answer_id": 5,
        "body": "Something pretty durable but I can't be sure",
        "date": "2018-01-04T00:00:00.000Z",
        "answerer_name": "metslover",
        "helpfulness": 5,
        "photos": [{
            "id": 1,
            "url": "urlplaceholder/answer_5_photo_number_1.jpg"
          },
          {
            "id": 2,
            "url": "urlplaceholder/answer_5_photo_number_2.jpg"
          },
        ]
      }
    ]
  }

  module.exports = sampleProductQuestions;