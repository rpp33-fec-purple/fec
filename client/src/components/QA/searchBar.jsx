import React from 'react';
import styled from 'styled-components';

const Container = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  // height: 70%;
  // box-sizing: border-box;
  // margin: 0;
  // padding: 5px;
  padding: 10px 32px;

  // font-family: 'Arial', sans-serif;
`;

const StyledSearchBar = styled.div`
  body{
    background: #f2f2f2;
    font-family: 'Arial', sans-serif;
  }

  .search {
    width: 100%;
    position: relative;
    display: flex;
  }

  .searchTerm {
    width: 100%;
    border: 3px solid #724060;
    padding: 5px;
    height: 35px;
    // width: 100px;
    outline: none;
    color: #724060;
    font-size: 16px;

    // border-right: none;
    // border-radius: 5px 0 0 5px;
  }

  .searchTerm:focus{
    color: #724060;
  }

  .searchButton {
    width: 50px;
    height: 25px;
    border: 1px solid #724060;
    background: #724060;
    text-align: center;
    color: #fff;
    border-radius: 0 5px 5px 0;
    cursor: pointer;
    font-size: 13px;
    padding-left: 2px;
    padding-right: 2px;
  }

  /*Resize the wrap to see the search bar change!*/
  .wrap{
    // width: 30%;
    // position: relative;
    // top: 50%;
    // left: 50%;
    // transform: translate(-50%, -50%);
  }


`;

class SearchBar extends React.Component {
  constructor(props) {
    super(props);
    this.state= {
      searchTerm: ''
    }

    this.handleChange = this.handleChange.bind(this);
    this.sortWithFilter = this.sortWithFilter.bind(this);

  }

  handleChange(e) {
    e.preventDefault();
    var term = e.target.value;
    if (term.length >= 3) {
      this.setState({
        searchTerm: e.target.value
      }, () => {
        this.sortWithFilter(this.state.searchTerm)
      })
    } else {
      this.setState({
        searchTerm: ''
      }, () => {
        this.sortWithFilter(this.state.searchTerm)
      })
    }
  }

  sortWithFilter() {
    var term = this.state.searchTerm;
    $.ajax({
      url: `${baseUrl}/qa/questions`,
      data: {
        product_id: this.props.productID,
        page: 1,
        count: 100
      },
      method: 'GET',
      success: (data) => {
        // console.log('data in client', data);
        var filteredList = data.results.filter((val) => {
          if (term === '') {
            return val;
          } else if (val.question_body.toLowerCase().includes(term.toLowerCase())) {
            return val;
          }
        })
        // console.log('filtered list', filteredList)
        if (term === '') {
          this.props.updateStateCauseFilter(filteredList, false, term)
        } else {
          this.props.updateStateCauseFilter(filteredList, true, term)
        }
      },
      error: (err) => {
        console.log('Error with GET request in filter:', err);
      }
    });

    // var term = this.state.searchTerm;
    // var filteredList = this.props.qAndAList.filter((val) => {
    //   if (term === '') {
    //     return val;
    //   } else if (val.question_body.toLowerCase().includes(term.toLowerCase())) {
    //     return val;
    //   }
    // })
    // console.log('filtered list', filteredList)
    // if (term === '') {
    //   this.props.updateStateCauseFilter(filteredList, false, term)
    // } else {
    //   this.props.updateStateCauseFilter(filteredList, true, term)
    // }
  }

  render() {
    return (
      <Container>
        <StyledSearchBar>
          <div id='qaSearchBar'>
            <form className='search'>
              <label htmlFor='question' id='qaSearchBar'></label>
              <input type='text' name='question' className='searchTerm' size="97" placeholder='HAVE A QUESTION? SEARCH FOR ANSWERS...' onChange={this.handleChange}></input>
              {/* <button type="submit" className='searchButton'>Search</button> */}
            </form>
          </div>
        </StyledSearchBar>
      </Container>
    )
  }
}

export default SearchBar;
