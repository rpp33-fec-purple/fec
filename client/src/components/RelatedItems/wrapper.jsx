import React from 'react';

class RelatedItems extends React.Component {
  constructor(props) {
    super(props);
  }
  // {this.props.data}
  render() {
    return (
      <h2>Related Items and Outfit Component</h2>
      // < RelatedItems data = {this.state.items} --> connect(mapStateToProps)(RelatedItems)
    )
  }
}

export default RelatedItems;
/*
 mapStateToProps = (state) => {
 return {
   data: state.items
 }
};





*/