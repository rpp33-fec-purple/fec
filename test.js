import React from 'react';
import TestRenderer from 'react-test-renderer';
import RelatedItems from './client/src/components/RelatedItems/wrapper';
import QA from './client/src/components/qa/wrapper';

describe('render the component RelatedItems', () => {
const testRenderer = TestRenderer.create(<RelatedItems />);
const testInst = testRenderer.root;
it('should contain a h2 element ', () => {
  expect(testRenderer.toJSON().type).toBe('h2');
})

// it('should return the right inner text of the existing element ', () => {
//   expect(testInst.findByProps({className:'Related Items'}).children[0]).toEqual('Related Items and Outfit Component')
// })

// it('the component should match the saved snapshot', () => {
//   expect(testRenderer.toJSON()).toMatchSnapshot();
// })

})

// it('render the component Q&A', () => {
//   const testRenderer = TestRenderer.create(<QA />);
//   expect(testRenderer.toJSON().children[0]).toEqual('Questions and Answers Component')
//   expect(testRenderer.toJSON()).toMatchSnapshot();
// })