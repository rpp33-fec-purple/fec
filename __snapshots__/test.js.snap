// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render the component Q&A 1`] = `
<h2
  className="QA"
>
  Questions and Answers Component
</h2>
`;
